### DISCUSSION TITLE
list files watched ?

### DISCUSSION THREAD

**Discussion by simonmichael** [watchexec/watchexec#731] [Questions & Help] [answered]
Thank you for watchexec !!

One thing I want to do with every watcher is list the files watched, for troubleshooting and for understanding the effect of options, VCS's, ignore files. But I could not find a way to do it with watchexec.

**Comment by passcod** [comment, marked as answer]
There's several levels at which you might look:
- paths given to the notify crate. That's logged at multiple levels but is generally either the "project origin" or the current directory if running without `-w` options, or the exact set of paths given by `-w` options.
- actual paths watched at kernel level. Unfortunately this is internal to notify and varies between OSes. For example on Linux notify recurses to resolve all subdirectories, while on Mac the kernel API itself recurses internally, and on BSD iirc notify recurses to resolve all subdirectories and files. Depending on the notify backend, it may be logged; in the CLI it will be visible at `-vvvv` level, or with `RUST_LOG=notify=trace`.
- events received from notify pre-filtering. This is available in logs. Run watchexec with the `RUST_LOG=watchexec::sources::fs=trace` env var and look for the message `"receiving possible event from watcher"`.
- events after filtering. There's `--print-events` for that.

You can also disable all implicit filtering with `--ignore-nothing` and use `--print-events`.

One other significant source of obscurity when looking into filters is that underlying matching based on glob patterns is opaque, there's no way to trace which patterns are applying in which order, and what's rejecting a particular path. This is planned to be improved on later.
